File: tagview/src/main/java/nesto/tagview/TagView.java
package nesto.tagview;

import android.content.Context;
import android.support.annotation.ColorInt;
import android.support.annotation.DrawableRes;
import android.support.annotation.IntRange;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.view.View;
import android.widget.FrameLayout;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * Created on 2016/11/3.
 * By nesto
 */

public class TagView extends FrameLayout
        implements OnTagClickListener,
        OnTagLongClickListener {
    private RecyclerView recyclerView;
    private List<Tag> tags;
    private TextPaint paint;
    private Context context;
    private TagAdapter adapter;
    private GridLayoutManager layoutManager;
    private int width;
    private OnTagClickListener clickListener;
    private OnTagLongClickListener longClickListener;

    private static final int ADDITION_MARGIN = 2;
    private static final int ADDITION_PADDING = 16;
    private int padding = ADDITION_PADDING;
    private int margin = ADDITION_MARGIN;
    // 2 * margin + 2 * padding
    private Integer additionTotal;

    private static final int DEFAULT_TEXT_SIZE = 12;
    private int textSize = DEFAULT_TEXT_SIZE;

    public TagView(Context context) {
        super(context);
        initView(context);
    }

    public TagView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initView(context);
    }

    private void initView(Context context) {
        this.context = context;
        View view = inflate(context, R.layout.tag_view, null);
        addView(view);
        layoutManager = new GridLayoutManager(context, 1);
        recyclerView = (RecyclerView) view.findViewById(R.id.recyclerView);
        tags = new ArrayList<>();
        adapter = new TagAdapter(context, tags);
        recyclerView.setAdapter(adapter);
        additionTotal = calculateAddition();
        initTextPaint();
    }

    private int calculateAddition() {
        return DpTrans.dp2px(context, 2 * margin + 2 * padding);
    }

    private void initTextPaint() {
        paint = new TextPaint();
        paint.setTextSize(DpTrans.dp2px(context, textSize));
    }

    private void setTagSize() {
        layoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override public int getSpanSize(int position) {
                String text = tags.get(position).tag;
                int size = (int) paint.measureText(text) + additionTotal;
                return Math.min(size, width);
            }
        });
    }

    public TagView textSize(@IntRange(from = 1) int dpValue) {
        textSize = dpValue;
        paint.setTextSize(DpTrans.dp2px(context, textSize));
        adapter.setTextSize(dpValue);
        setTagSize();
        return this;
    }

    public TagView textColor(@ColorInt int color) {
        adapter.setTextColor(color);
        return this;
    }

    public TagView backgroundColor(@ColorInt int color) {
        adapter.setBackground(color);
        return this;
    }

    public TagView backgroundDrawable(@DrawableRes int drawable) {
        adapter.setBackgroundDrawable(drawable);
        return this;
    }

    public TagView padding(@IntRange(from = 0) int padding) {
        this.padding = Math.max(padding, 8);
        additionTotal = calculateAddition();
        setTagSize();
        adapter.setPadding(DpTrans.dp2px(context, padding));
        return this;
    }

    public TagView margin(@IntRange(from = 0) int margin) {
        this.margin = margin;
        additionTotal = calculateAddition();
        setTagSize();
        adapter.setMargin(DpTrans.dp2px(context, margin));
        return this;
    }

    public TagView itemHeight(@IntRange(from = 1) int height) {
        adapter.setHeight(DpTrans.dp2px(context, height));
        return this;
    }

    public TagView corner(@IntRange(from = 0) int radius) {
        adapter.setRadius(DpTrans.dp2px(context, radius));
        return this;
    }

    public TagView dividerHeight(@IntRange(from = 0) int height) {
        adapter.setDivider(DpTrans.dp2px(context, height));
        return this;
    }


    public TagView addStringTags(Collection<String> tagCollection) {
        int start = tags.size();
        for (String s : tagCollection) {
            tags.add(new Tag(s, null, null));
        }
        adapter.notifyItemRangeInserted(start, tagCollection.size());
        return this;
    }

    public TagView addStringTag(String tag) {
        tags.add(new Tag(tag, null, null));
        adapter.notifyItemInserted(tags.size() - 1);
        return this;
    }

    public TagView addTags(Collection<Tag> tagCollection) {
        int start = tags.size();
        tags.addAll(tagCollection);
        adapter.notifyItemRangeInserted(start, tagCollection.size());
        return this;
    }

    public TagView addTag(Tag tag) {
        tags.add(tag);
        adapter.notifyItemInserted(tags.size() - 1);
        return this;
    }

    public TagView setListener(OnTagClickListener listener) {
        this.clickListener = listener;
        adapter.setClickListener(listener);
        return this;
    }

    public TagView setListener(OnTagLongClickListener listener) {
        this.longClickListener = listener;
        adapter.setLongClickListener(listener);
        return this;
    }

    @Override public void tagClicked(int position, String item) {
        if (clickListener != null) clickListener.tagClicked(position, item);
    }

    @Override public void tagLongClicked(int position, String item) {
        if (longClickListener != null) longClickListener.tagLongClicked(position, item);
    }

    @Override protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        width = recyclerView.getMeasuredWidth() - recyclerView.getPaddingLeft()
                - recyclerView.getPaddingRight();
        // make sure the width is at least 1
        width = Math.max(1, width);
        layoutManager.setSpanCount(width);
        setTagSize();
        recyclerView.setLayoutManager(layoutManager);
    }

    public TagView removeAllTags() {
        int count = tags.size();
        tags.clear();
        adapter.notifyItemRangeRemoved(0, count);
        return this;
    }

    public TagView removeTag(int position) {
        tags.remove(position);
        adapter.notifyItemRemoved(position);
        return this;
    }
}

File: tagview/src/main/java/nesto/tagview/TagView.kt
package nesto.tagview

import android.content.Context
import android.support.annotation.ColorInt
import android.support.annotation.DrawableRes
import android.support.annotation.IntRange
import android.support.v7.widget.GridLayoutManager
import android.text.TextPaint
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import kotlinx.android.synthetic.main.tag_view.view.*

/**
 * Created on 2016/11/3.
 * By nesto
 */

class TagView : FrameLayout {

    companion object {
        private val ADDITION_MARGIN = 2
        private val ADDITION_PADDING = 16

        private val DEFAULT_TEXT_SIZE = 12
    }

    private lateinit var tags: MutableList<Tag>
    private lateinit var paint: TextPaint
    private lateinit var adapter: TagAdapter
    private lateinit var layoutManager: GridLayoutManager
    private var spanWidth: Int = 0
    private var clickListener: ((Int, String) -> Unit)? = null
    private var longClickListener: ((Int, String) -> Unit)? = null
    private var padding = ADDITION_PADDING
    private var margin = ADDITION_MARGIN
    // 2 * margin + 2 * padding
    private var additionTotal: Int = 0

    private var textSize = DEFAULT_TEXT_SIZE

    constructor(context: Context) : super(context) {
        initView(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        initView(context)
    }

    private fun initView(context: Context) {
        val view = View.inflate(context, R.layout.tag_view, null)
        addView(view)
        layoutManager = GridLayoutManager(context, 1)
        tags = mutableListOf<Tag>()
        adapter = TagAdapter(context, tags)
        recyclerView.adapter = adapter
        additionTotal = calculateAddition()
        initTextPaint()
    }

    private fun calculateAddition(): Int {
        return dp2px(context!!, (2 * margin + 2 * padding).toFloat())
    }

    private fun initTextPaint() {
        paint = TextPaint()
        paint.textSize = dp2px(context!!, textSize.toFloat()).toFloat()
    }

    private fun setTagSize() {
        layoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                val text = tags[position].tag
                val size = paint.measureText(text).toInt() + additionTotal
                return Math.min(size, spanWidth)
            }
        }
    }

    fun textSize(@IntRange(from = 1) dpValue: Int): TagView {
        textSize = dpValue
        paint.textSize = dp2px(context, textSize.toFloat()).toFloat()
        adapter.setTextSize(dpValue)
        setTagSize()
        return this
    }

    fun textColor(@ColorInt color: Int): TagView {
        adapter.setTextColor(color)
        return this
    }

    fun backgroundColor(@ColorInt color: Int): TagView {
        adapter.setBackground(color)
        return this
    }

    fun backgroundDrawable(@DrawableRes drawable: Int): TagView {
        adapter.setBackgroundDrawable(drawable)
        return this
    }

    fun padding(@IntRange(from = 0) padding: Int): TagView {
        this.padding = Math.max(padding, 8)
        additionTotal = calculateAddition()
        setTagSize()
        adapter.setPadding(dp2px(context, padding.toFloat()))
        return this
    }

    fun margin(@IntRange(from = 0) margin: Int): TagView {
        this.margin = margin
        additionTotal = calculateAddition()
        setTagSize()
        adapter.setMargin(dp2px(context, margin.toFloat()))
        return this
    }

    fun itemHeight(@IntRange(from = 1) height: Int): TagView {
        adapter.setHeight(dp2px(context, height.toFloat()))
        return this
    }

    fun corner(@IntRange(from = 0) radius: Int): TagView {
        adapter.setRadius(dp2px(context!!, radius.toFloat()))
        return this
    }


    fun dividerHeight(@IntRange(from = 0) height: Int): TagView {
        adapter.setDivider(dp2px(context!!, height.toFloat()))
        return this
    }

    fun addStringTags(tagCollection: Collection<String>): TagView {
        val start = tags.size
        for (s in tagCollection) {
            tags.add(Tag(s, null, null))
        }
        adapter.notifyItemRangeInserted(start, tagCollection.size)
        return this
    }

    fun addStringTag(tag: String, position: Int? = null): TagView {
        if (position == null) {
            tags.add(Tag(tag, null, null))
        } else {
            tags.add(position, Tag(tag, null, null))
        }
        adapter.notifyItemInserted(tags.size - 1)
        return this
    }

    fun addTags(tagCollection: Collection<Tag>): TagView {
        val start = tags.size
        tags.addAll(tagCollection)
        adapter.notifyItemRangeInserted(start, tagCollection.size)
        return this
    }

    fun addTag(tag: Tag, position: Int? = null): TagView {
        if (position == null) {
            tags.add(tag)
        } else {
            tags.add(position, tag)
        }
        adapter.notifyItemInserted(tags.size - 1)
        return this
    }

    fun setOnTagClickListener(listener: (Int, String) -> Unit): TagView {
        clickListener = listener
        adapter.setClickListener(listener)
        return this
    }

    fun setOnTagLongClickListener(listener: (Int, String) -> Unit): TagView {
        longClickListener = listener
        adapter.setLongClickListener(listener)
        return this
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        spanWidth = recyclerView.measuredWidth - recyclerView.paddingLeft
        -recyclerView.paddingRight
        // make sure the width is at least 1
        spanWidth = Math.max(1, spanWidth)
        layoutManager.spanCount = spanWidth
        setTagSize()
        recyclerView.layoutManager = layoutManager
    }

    fun removeAllTags(): TagView {
        val count = tags.size
        tags.clear()
        adapter.notifyItemRangeRemoved(0, count)
        return this
    }

    fun removeTag(position: Int): TagView {
        tags.removeAt(position)
        adapter.notifyItemRemoved(position)
        return this
    }
}
